server/test/unit/paTest/handleMessageClientsTest.js

```diff
@@ -0,0 +1,268 @@
+    fromPhoneNumber: testFromPhoneNumber,
+  },
+  {
+    id: 'client-bad-from-phone-number',
+    displayName: 'Client Bad From Phone Number',
+    responderPhoneNumbers: [failPhoneNumbers[0]],
+    fallbackPhoneNumbers: [],
+    heartbeatPhoneNumbers: [],
+    fromPhoneNumber: testBadFromPhoneNumber,
+  },
+]
+
+async function mockSendTwilioMessage(toPhoneNumber, fromPhoneNumber, twilioMessage) {
+  if (fromPhoneNumber === testBadFromPhoneNumber) {
+    return { status: 'error' }
+  }
+
+  // twilioHelpers.sendTwilioMessage does:
+  // helpers.log(`Sent by Twilio: ${response.sid}`)
+  // where 'response' comes from a Twilio API call.
+  helpers.log(`Sent by (mock) Twilio: ${twilioMessage}`)
+  return { status: 'queued' }
+}
+
+use(sinonChai)
+
+const sandbox = sinon.createSandbox()
+
+describe('pa.js unit tests: handleMessageClients', () => {
+  beforeEach(() => {
+    sandbox.stub(db, 'getActiveClients').returns(activeClients)
+  })
+
+  afterEach(() => {
+    sandbox.restore()
+  })
+
+  describe('for a request when Twilio is operating as expected', () => {
+    beforeEach(async () => {
+      sandbox.stub(twilioHelpers, 'sendTwilioMessage').callsFake(mockSendTwilioMessage)
+
+      this.res = mockResponse(sandbox)
+      await pa.handleMessageClients(
+        {
+          body: { twilioMessage: testTwilioMessage, googleIdToken: testGoogleIdToken },
+        },
+        this.res,
+      )
+    })
+
+    it('should respond with status 200', () => {
+      expect(this.res.status).to.be.calledWith(200)
+    })
+
+    it('should respond with the Twilio message as twilioMessage', () => {
+      expect(this.res.body.twilioMessage).to.equal(testTwilioMessage)
+    })
+
+    it('should respond with the expected phone numbers in successfullyMessaged', () => {
+      const toPhoneNumbers = this.res.body.successfullyMessaged.map(twilioTraceObject => {
+        return twilioTraceObject.to
+      })
+
+      // deep equal; no other phone numbers should be contacted
+      expect(toPhoneNumbers).to.eql(successfulPhoneNumbers)
+    })
+
+    it('should respond with the expected phone numbers in failedToMessage', () => {
+      const toPhoneNumbers = this.res.body.failedToMessage.map(twilioTraceObject => {
+        return twilioTraceObject.to
+      })
+
+      // deep equal; no other phone numbers should fail
+      expect(toPhoneNumbers).to.eql(failPhoneNumbers)
+    })
+
+    it('should call twilioHelpers.sendTwilioMessage for all active clients', () => {
+      const expectedArguments = []
+
+      activeClients.forEach(client => {
+        // gather all phone numbers for this client
+        const phoneNumbers = []
+        phoneNumbers.push(...client.responderPhoneNumbers, ...client.fallbackPhoneNumbers, ...client.heartbeatPhoneNumbers)
+
+        // gather all unique phone numbers for this client
+        const uniquePhoneNumbers = new Set()
+        phoneNumbers.forEach(phoneNumber => {
+          uniquePhoneNumbers.add(phoneNumber)
+        })
```

Comment:
I see that you've used the same code in your test as in the production code that you're testing. So if there is a bug in this implementation, you won't find it. Can you create `expectedArguments` in a different way here?

Reply:
I can't make it different, unfortunately, unless this code was exported as a helper function from pa.js, or in brave-alert-lib. But then I feel I ought to have test cases for that function, too, and so on... :rofl:

Reply:
Could you hardcode it, considering you know what clients and phone numbers you've used?

Reply:
True, I could! I can add this at the top of the file, then. :)

Reply:
Done!